docs/source/architecture/journal.rst

```diff
@@ -0,0 +1,476 @@
+decommit. These lists are the blocks in each fork back to the common root.
+
+The BlockCache
+==============
+
+The Block Cache holds the working set of blocks for the validator and tracks the
+processing state. This processing state is tracked as valid, invalid, or
+unknown. Valid blocks are blocks that have been proven to be valid by the
+ChainController. Invalid blocks are blocks that failed validation or have an
+invalid block as a predecessor. Unknown are blocks that have not yet completed
+validation, usually having just arrived from the Completer.
+
+The BlockCache is an in-memory construct. It is rebuilt by demand when the
+system is started.
+
+If a block is not present in the BlockCache, it will look in the BlockStore for
+the block. If it is not found or the lookup fails, the block is unknown to the
+system. If the block is found in the BlockStore it is loaded into the
+BlockCache and marked as valid. All blocks in the BlockStore are considered
+valid.
+
+The BlockCache keeps blocks that are currently relevant, tracked by the last
+time the block was accessed. Periodically, the blocks that have not been
+accessed recently are purged from the block cache. The time horizon for purging
+and the frequency of purging are either configured or determined automatically
+based on the average block time of the system.
+
+The Completer 
+=============
+
+The Completer is responsible for making sure Blocks and Batches are complete
+before they are delivered. Blocks are considered formally complete once all of
+their predecessors have been delivered to the ChainController and their batches
+field contains all the Batches specified in the BlockHeader’s batch_ids list.
+The batches field is also expected to be in the same order as the batch_ids.
+Once Blocks are formally complete they are delivered to the ChainController for
+validation.
+
+Each batch is considered complete once all of its dependent transactions (from
+earlier blocks or batches) have been processed (either by the ChainController
+or the BlockPublisher respectively).
+
+All Blocks and Batches will have a timeout for being completed. After the
+initial request for the missing dependencies is sent, if the response is not
+received within the specified time window, they are dropped.
+
+If you have a new block of unknown validity, you must ensure that its
+predecessors have been delivered to the journal. If a predecessor is not
+delivered on request to the journal in a reasonable amount of time, the new
+block cannot be validated.
+
+Consider the case where you have the chain A->B->C :
+
+If C arrives and B is not in the BlockCache, the validator will request B. If
+the request for B times out, the C block is dropped.
+
+If later on D arrives with predecessor C, of chain A->B->C->D, the Completer
+will request C from the network and once C arrives, then will request B again.
+If B arrives this time then the chain will be validated.
```

Comment:
If B arrives this time then the new chain will be delivered to the ChainController, where they will be check for validity and considered for becoming the block head by the ChainController.